geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/data/CassandraFeatureStore.scala

```diff
@@ -0,0 +1,177 @@
+    else                                   new UpdateFW(contentState.sft, contentState.session)
+  }
+
+  override def buildFeatureType(): SimpleFeatureType = contentState.sft
+
+  override def getBoundsInternal(query: Query): ReferencedEnvelope = new ReferencedEnvelope(-180.0, 180.0, -90.0, 90.0, DefaultGeographicCRS.WGS84)
+
+  override def getCountInternal(query: Query): Int = {
+    // TODO: might overflow
+    if(query.equals(Query.ALL) || FilterHelper.isFilterWholeWorld(query.getFilter)) {
+      contentState.session.execute(contentState.ALL_COUNT_QUERY.bind()).iterator().next().getLong(0).toInt
+    } else {
+      val plans = planQuery(query)
+      executeGeoTimeCountQuery(query, plans).toInt
+    }
+  }
+
+  case class RowAndColumnQueryPlan(row: Int, lz3: Long, uz3: Long, contained: Boolean)
+
+  def executeGeoTimeCountQuery(query: Query, plans: Seq[RowAndColumnQueryPlan]) = {
+    // TODO: currently overestimates the count in order to increase performance
+    // Fix overestimation by pushing geo etc predicates down into the database
+    val features = contentState.builderPool.withResource { builder =>
+      val futures = plans.map { case RowAndColumnQueryPlan(r, l, u, contained) =>
+        val q = contentState.GEO_TIME_COUNT_QUERY.bind(r: Integer, l: lang.Long, u: lang.Long)
+        contentState.session.executeAsync(q)
+      }
+      futures.flatMap { f => f.get().iterator().toList }.map(_.getLong(0)).sum
+    }
+    features
+  }
+
+
+  override def getReaderInternal(query: Query): FeatureReader[SimpleFeatureType, SimpleFeature] = {
+    val iter =
+      if(query.equals(Query.ALL) || FilterHelper.isFilterWholeWorld(query.getFilter)) {
+        getAllFeatures
+      } else {
+        val plans    = planQuery(query)
+        val features = executeGeoTimeQuery(query, plans)
+        features.iterator
+      }
+    new DelegateSimpleFeatureReader(contentState.sft, new DelegateSimpleFeatureIterator(iter))
+  }
+
+  def executeGeoTimeQuery(query: Query, plans: Seq[RowAndColumnQueryPlan]): Seq[SimpleFeature] = {
+    val features = contentState.builderPool.withResource { builder =>
+      val futures = plans.map { case RowAndColumnQueryPlan(r, l, u, contained) =>
+        val q = contentState.GEO_TIME_QUERY.bind(r: Integer, l: lang.Long, u: lang.Long)
+        (contained, contentState.session.executeAsync(q))
+      }
+      futures.flatMap { case (contains, fut) =>
+        postProcessResults(query, builder, contains, fut)
+      }
+    }
+    features
+  }
+
+  val WHOLE_WORLD = new ReferencedEnvelope(-180.0, 180.0, -90.0, 90.0, DefaultGeographicCRS.WGS84)
+  def planQuery(query: Query) = {
+    import org.locationtech.geomesa.filter._
+    import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType._
+
+    val origBounds = query.getFilter.accept(ExtractBoundsFilterVisitor.BOUNDS_VISITOR, DefaultGeographicCRS.WGS84).asInstanceOf[Envelope]
```

Comment:
This code is perfectly fine as-is; I just wanted to add a comment for our internal edification...

This GeoTools code should be considered a "fast, but pessimistic" routine.  We could tighten the bounding Geometry, albeit at the cost of some processing time (and how complex a return geometry we wish to entertain).  We used to do this, but removed it during a refactor.